### DISCUSSION TITLE
gasPrice miscalulation when using celo feeCurrency

### DISCUSSION THREAD

**Discussion by aaronmgdr** [wevm/viem#914] [Idea / Feature Request]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

all

### Current Behavior

currently when sending tx that sets feeCurrency the price of gas is set using the chains native token (CELO) based on block headers. 

this kinda works if CELO is worth less than the feeCurrency (as is the case now) but if fee 1 unit of fee currency is worth more than 1 unit of CELO it will fail) 

### Expected Behavior

when the `tx` has `feeCurrency` set then need to set `tx.maxFeePerGas` to `client.request({method: 'eth_gasPrice', params [tx.feeCurrency]}`  in `prepareRequest`

### Steps To Reproduce

send a tx on a celo chain with `feeCurrency` set. 

### Anything else?

This is on me for not considering this sooner. 

my current implementation thought is to have a chain specific preparer that gets called  towards the top of `prepareRequest`

something like 
* in existing prepareRequest function*
```ts
const block = await getBlock(client, { blockTag: 'latest' })
/// new code
const chainPreparedRequest = client.chain.prepareRequest(client, args)
  const request = { ...args, ...chainPreparedRequest, from: account.address }

```

*new function in chains/celo*
```ts

async prepareRequest(client, tx) {
const prepared = {...tx}
if (tx.feeCurrency) {
 prepared.gasPrice = client.request({method: 'eth_gasPrice', params [tx.feeCurrency]}
}

return prepared
}

```

**Comment by jxom** [comment]
Yep! We are planning to add `getFees` or alike as a chain-derived function to viem which will handle this. I'm back from vacation on Aug 18, but this is one of my first things on my TODO list when I get back. Going to move this as a discussion as it is more a feature request to improve DX. Folks can derive this manually (like in your example), until a built-in solution is provided.

**Reply by aaronmgdr** [reply]
have nice holiday

**Reply by Jesiiia** [reply]
Yea i mean with you

**Reply by evseevnn** [reply]
Any update?

**Reply by jxom** [reply]
https://viem.sh/docs/clients/chains.html#fees-estimatefeespergas

**Comment by aaronmgdr** [comment]
@jxom , Im finally working on special case for celo of getting the feePerGas when fee is paid in celo feecurrency. 

Ideally id like to only use a custom estimateFeePerGas when its needed and otherwise just use the default estimateFeePerGas however since that will call the chain.fees.estimateFeePerGas if it exists ive ran into infinite loop. 
 
My thought right now is to add an additional param to skip the chain fee estimator

**Reply by aaronmgdr** [reply]
The spoken of [param to skip using chain estimator ](https://github.com/celo-org/viem/pull/10/files#diff-adc7f2d066442058dfa888456e28d1033d1773770bce5da2c4bf2b1c2c639363)
